src/camellia_fmt_plug.c

```diff
@@ -0,0 +1,247 @@
+	{"$camellia$NeEGbM0Vhz7u+FGJZrcPiw==", "admin" },
+	{"$camellia$ItGoyeyQIvPjT/qBoDKQZg==", "pswpsw" },
+	{NULL}
+};
+
+static char (*saved_key)[PLAINTEXT_LENGTH + 1];
+static ARCH_WORD_32 (*crypt_out)[BINARY_SIZE / sizeof(ARCH_WORD_32)];
+
+static void init(struct fmt_main *self)
+{
+#ifdef _OPENMP
+	int omp_t = omp_get_max_threads();
+	self->params.min_keys_per_crypt *= omp_t;
+	omp_t *= OMP_SCALE;
+	self->params.max_keys_per_crypt *= omp_t;
+#endif
+	saved_key = mem_calloc(self->params.max_keys_per_crypt,
+	                       sizeof(*saved_key));
+	crypt_out = mem_calloc(self->params.max_keys_per_crypt,
+	                       sizeof(*crypt_out));
+}
+
+static void done(void)
+{
+	MEM_FREE(crypt_out);
+	MEM_FREE(saved_key);
+}
+
+static int valid(char *ciphertext, struct fmt_main *self)
+{
+	char *p;
+
+	if (strncmp(ciphertext, FORMAT_TAG, TAG_LENGTH))
+		return 0;
+
+	p = ciphertext + TAG_LENGTH;
+	if (strlen(p) != BINARY_ENCODED_SIZE)
+		return 0;
+
+	if (BINARY_ENCODED_SIZE - 2!= base64_valid_length(p, e_b64_mime, flg_Base64_MIME_TRAIL_EQ))
+                return 0;
+
+	return 1;
+}
+
+static void *get_binary(char *ciphertext)
+{
+	static union {
+		unsigned char c[BINARY_SIZE + 1 + 4];
+		ARCH_WORD dummy;
+	} buf;
+	unsigned char *out = buf.c;
+	char *p;
+	p = strrchr(ciphertext, '$') + 1;
+	base64_decode((char*)p, BINARY_ENCODED_SIZE, (char*)out);
+
+	return out;
+}
+
+static int get_hash_0(int index) { return crypt_out[index][0] & PH_MASK_0; }
+static int get_hash_1(int index) { return crypt_out[index][0] & PH_MASK_1; }
+static int get_hash_2(int index) { return crypt_out[index][0] & PH_MASK_2; }
+static int get_hash_3(int index) { return crypt_out[index][0] & PH_MASK_3; }
+static int get_hash_4(int index) { return crypt_out[index][0] & PH_MASK_4; }
+static int get_hash_5(int index) { return crypt_out[index][0] & PH_MASK_5; }
+static int get_hash_6(int index) { return crypt_out[index][0] & PH_MASK_6; }
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index = 0;
+#ifdef _OPENMP
+#pragma omp parallel for
+#endif
+	for (index = 0; index < count; index++)
+	{
+		CAMELLIA_KEY st_key;
+		unsigned char in[16] = {0};
+		unsigned char key[32] = {0};
+		memcpy(key, saved_key[index], strlen(saved_key[index]));
+		Camellia_set_key(key, 256, &st_key);
+		Camellia_encrypt(in, (unsigned char*)crypt_out[index], &st_key);
+	}
+
+	return count;
+}
+
+static int cmp_all(void *binary, int count)
+{
+	int index = 0;
+#ifdef _OPENMP
+	for (; index < count; index++)
+#endif
+		if (!memcmp(binary, crypt_out[index], ARCH_SIZE))
+			return 1;
+	return 0;
+}
+
+static int cmp_one(void *binary, int index)
+{
+	return !memcmp(binary, crypt_out[index], BINARY_SIZE);
+}
+
+static int cmp_exact(char *source, int index)
+{
+	return 1;
+}
+
+static void camellia_set_key(char *key, int index)
+{
+	int len = strlen(key);
+
+	memcpy(saved_key[index], key, len);
+	saved_key[index][len] = 0;
+}
```

Comment:
What's the point of emulating strcpy like this? It could be just
```c
static void camellia_set_key(char *key, int index)
{
	strnzcpy(saved_key[index], key, PLAINTEXT_LENGTH + 1);
}
```